File: lib/kwartz/translator.rb
####
#### $Rev$
#### $Release$
#### $Copyright$
####


require 'kwartz/node'
require 'abstract'



module Kwartz


  ##
  ## .[abstract] translate list of Statement into target code (eRuby, PHP, ...)
  ##
  class Translator


    ## .[abstract] translate list of Statement into String and return it
    def translate(stmt_list)
      not_implemented
    end


    ## .[abstract] translate NativeStatement using visitor pattern
    def translate_native_stmt(stmt)
      not_implemented
    end


    ## .[abstract] translate PrintStatement using visitor pattern
    def translate_print_stmt(stmt)
      not_implemented
    end


    ## .[abstract] translate NativeExpression using visitor pattern
    def translate_native_expr(expr)
      not_implemented
    end


    ## .[abstract] translate String using visitor pattern
    def translate_string(str)
      not_implemented
    end


#    ## [abstract] translate TextExpression using visitor pattern
#    def translate_text_expr(expr)
#      not_implemented
#    end


    @@class_table = {}


    def self.register_class(lang, klass)
      @@class_table[lang] = klass
    end


    def self.get_class(lang)
      return @@class_table[lang]
    end


  end #class



  ##
  ## concrete class for visitor pattern
  ##
  ## see ErbTranslator, PhpTranslator, JstlTranslator, and so on for detail.
  ##
  class BaseTranslator < Translator


    def initialize(marks=[], properties={})
      @stmt_l, @stmt_r, @expr_l, @expr_r, @escape_l, @escape_r = marks
      @nl = properties[:nl] || "\n"
      @escape = properties[:escape]
      @escape = Config::PROPERTY_ESCAPE if @escape == nil
      @header = properties[:header]
      @footer = properties[:footer]
    end


    attr_accessor :escape, :header, :footer


    def translate(stmt_list)
      @sb = ''
      @sb << @header if @header
      stmt_list.each do |stmt|
        stmt.accept(self)
      end
      @sb << @footer if @footer
      return @sb
    end


    def translate_native_stmt(stmt)
      @sb << @stmt_l << stmt.code << @stmt_r   # ex. <% stmt.code %>
      @sb << @nl unless stmt.no_newline
    end


    def translate_print_stmt(stmt)
      stmt.args.each do |arg|
        #arg.accept(self)
        if arg.is_a?(String)
          #translate_string(arg)
          parse_embedded_expr(arg)
        elsif arg.is_a?(NativeExpression)
          translate_native_expr(arg)
        else
          assert
        end
      end
    end


    def translate_native_expr(expr)
      assert unless expr.is_a?(NativeExpression)
      flag_escape = expr.escape?
      flag_escape = @escape if flag_escape == nil
      if flag_escape
        add_escaped_expr(expr.code)
      else
        add_plain_expr(expr.code)
      end
    end


    def translate_string(str)
      @sb << str
    end


#    def translate_text_expr(expr)
#      @sb << expr.text
#    end


    protected


    def parse_embedded_expr(text)
      pos = 0
      text.scan(/@(!*)\{(.*?)\}@/) do |indicator, expr_str|
        m = Regexp.last_match
        s = text[pos, m.begin(0) - pos]
        pos = m.end(0)
        translate_string(s) unless s.empty?
        expr_str = parse_expr_str!(expr_str)
        case indicator
        when nil, ''  ;  add_escaped_expr(expr_str)
        when '!'      ;  add_plain_expr(expr_str)
        when '!!'     ;  add_debug_expr(expr_str)
        else          ;  # do nothing
        end
      end
      rest = pos == 0 ? text : $'
      translate_string(rest) unless rest.empty?
    end


    ## (abstract) convert expression to native expression string.
    ## ex. 'item.name' => '$item->name', '$item->name' => '$item->name'
    def parse_expr_str!(expr_str)
      not_implemented
    end


    def add_plain_expr(expr_code)
      @sb << @expr_l << expr_code << @expr_r       # ex. <%= expr_code %>
    end


    def add_escaped_expr(expr_code)
      @sb << @escape_l << expr_code << @escape_r   # ex. <%=h expr_code %>
    end


    def add_debug_expr(expr_code)
      not_implemented
    end


    ## concat several print statements into a statement
    def optimize_print_stmts(stmt_list)
      stmt_list2 = []
      args = []
      stmt_list.each do |stmt|
        if stmt.is_a?(PrintStatement)
          args += stmt.args
        else
          if !args.empty?
            args = _compact_args(args)
            stmt_list2 << PrintStatement.new(args)
            args = []
          end
          stmt_list2 << stmt
        end
      end
      if !args.empty?
        args = _compact_args(args)
        stmt_list2 << PrintStatement.new(args)
      end
      return stmt_list2
    end


    private


    ## concat several string arguments into a string in arguments
    def _compact_args(args)
      args2 = []
      s = ''
      args.each do |arg|
        if arg.is_a?(NativeExpression)
          args2 << s unless s.empty?
          s = ''
          args2 << arg
        else
          s << arg
        end
      end
      args2 << s unless s.empty?
      return args2
    end


  end #class



  module NoTextEnhancer  # :nodoc:


    def translate_string(str)
      @sb << (@nl * str.count("\n"))
      pos = str.rindex("\n")
      len = pos ? str.length - pos - 1 : str.length
      @sb << (' ' * len)
    end


  end



end #module

File: Kwartz/Translator.php
<?php
// vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4:

// $Rev$
// $Release$
// $Copyright$


require_once('Kwartz/Exception.php');
require_once('Kwartz/Node.php');
require_once('Kwartz/Utility.php');
require_once('Kwartz/Config.php');


/**
 *  translate list of Statement into target code (eRuby, PHP, ...)
 */
abstract class KwartzTranslator {

    /**
     *  translate list of Statement into String and return it
     */
    public function translate($stmt_list) {}
    //abstract public function translate($stmt_list);    // pear package error


    /**
     *  translate NativeStatement using visitor pattern
     */
    function translate_native_stmt($stmt) {}
    //abstract function translate_native_stmt($stmt);    // pear package error


    /**
     *  translate PrintStatement using visitor pattern
     */
    function translate_print_stmt($stmt) {}
    //abstract function translate_print_stmt($stmt);    // pear package error


    /**
     *  translate NativeExpression using visitor pattern
     */
    function translate_native_expr($expr) {}
    //abstract function translate_native_expr($expr);    // pear package error


    /**
     *  translate String using visitor pattern
     */
    function translate_string($str) {}
    //abstract function translate_string($str);    // pear package error

}



/**
 * concrete class for visitor pattern
 *
 * see ErbTranslator, PhpTranslator, JstlTranslator, and so on for detail.
 */
class KwartzBaseTranslator extends KwartzTranslator {


    var $stmt_l;
    var $stmt_r;
    var $expr_l;
    var $expr_r;
    var $escape_l;
    var $escape_r;
    var $nl;
    var $escape;
    var $header;
    var $footer;

    var $buf;

    function __construct($marks, $properties=array()) {
        list($this->stmt_l, $this->stmt_r, $this->expr_l, $this->expr_r,
             $this->escape_l, $this->escape_r) = $marks;
        $this->nl = kwartz_array_get($properties, 'nl', "\n");
        $this->escape = kwartz_array_get($properties, 'escape', KWARTZ_PROPERTY_ESCAPE);
        $this->header = kwartz_array_get($properties, 'header', null);
        $this->footer = kwartz_array_get($properties, 'footer', null);
    }


    function translate($stmt_list) {
        $this->buf = array();
        if ($this->header)
            $this->buf[] = $this->header;
        foreach ($stmt_list as $stmt) {
            $stmt->accept($this);
        }
        if ($this->footer)
            $this->buf[] = $this->footer;
        return join($this->buf);
    }


    function translate_native_stmt($stmt) {
        $this->buf[] = $this->stmt_l;   // ex. '<%' . $stmt->code . '%>'
        $this->buf[] = $stmt->code;
        $this->buf[] = $this->stmt_r;
        if (! $stmt->no_newline)
            $this->buf[] = $this->nl;
    }


    function translate_print_stmt($stmt) {
        foreach ($stmt->args as $arg) {
            if (is_string($arg)) {
                //$this->translate_string($arg);
                $this->parse_embedded_expr($arg);
            } else {
                assert('$arg instanceof KwartzNativeExpression');
                $this->translate_native_expr($arg);
            }
        }
    }


    function translate_native_expr($expr) {
        assert('$expr instanceof KwartzNativeExpression');
        $flag_escape = $expr->escape;
        if ($flag_escape === null) $flag_escape = $this->escape;
        if ($flag_escape) {
            $this->_add_escaped_expr($expr->code);
        } else {
            $this->_add_plain_expr($expr->code);
        }
    }


    function _add_plain_expr($expr_code) {
        // ex. "<\?php echo " . $expr->code . "; ?\>"
        $this->buf[] = $this->expr_l;
        $this->buf[] = $expr_code;
        $this->buf[] = $this->expr_r;
    }


    function _add_escaped_expr($expr_code) {
        // ex. "<\?php echo htmlspecialchars(".$expr->code)."; ?\>"
        $this->buf[] = $this->escape_l;
        $this->buf[] = $expr_code;
        $this->buf[] = $this->escape_r;
    }


    function _add_debug_expr($expr_code) {
        /* TBI */
    }


    function translate_string($str) {
        $this->buf[] = $str;
    }


    function parse_embedded_expr($text) {
        kwartz_scan_text('/@(!*)\{(.*?)\}@/', $text, $matched, $rest);
        foreach ($matched as $m) {
            $prev_text = $m[0];
            $indicator = $m[1];
            $expr_code = $m[2];
            $this->translate_string($prev_text);
            $len = strlen($indicator);
            switch ($len) {
            case 0:  $this->_add_escaped_expr($expr_code);  break;
            case 1:  $this->_add_plain_expr($expr_code);    break;
            case 1:  $this->_add_debug_expr($expr_code);    break;
            default:  // ignore
            }
        }
        if ($rest) {
            $this->translate_string($rest);
        }
    }



    // concat several print statements into a statement
    function optimize_print_stmts($stmt_list) {
        $stmt_list2 = array();   // list
        $args = array();  // list
        foreach ($stmt_list as $stmt) {
            if ($stmt instanceof KwartzPrintStatement) {
                kwartz_array_concat($args, $stmt->args);
            }
            else {
                if (count($args) != 0) {
                    $args = $this->_compact_args($args);
                    $stmt_list2[] = new KwartzPrintStatement($args);
                    $args = array();
                }
                $stmt_list2[] = $stmt;
            }
        }
        if (count($args) != 0) {
            $args = $this->_compact_args($args);
            $stmt_list2[] = new KwartzPrintStatement($args);
        }
        return $stmt_list2;
    }


    // concat several string arguments into a string in arguments
    function _compact_args(&$args) {
        $args2 = array();  // list
        $buf   = array();  // list
        foreach ($args as $arg) {
            if (is_string($arg)) {
                $buf[] = $arg;
            }
            else {
                assert('$arg instanceof KwartzNativeExpression');
                if (count($buf) > 0) {
                    $args2[] = join($buf);
                    $buf = array();
                }
                $args2[] = $arg;
            }
        }
        if (count($buf) > 0) {
            $args2[] = join($buf);
        }
        return $args2;
    }


}



?>
